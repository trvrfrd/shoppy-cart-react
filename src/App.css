* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  min-height: 100vh;
  display: flex;
  justify-content: center;
}

body * {
  border: 1px solid red;
}

#app {
  max-width: 1000px;
}

header {
  height: 100px;
  display: flex;
  justify-content: center;
  align-items: center;
}

main {
  padding: 20px 0;
  display: grid;
  height: calc(100% - 100px);
  grid-template-columns: 1fr 150px;
  grid-gap: 60px;
}

aside {
  margin-right: 20px;
}

.modal {
  align-items: center;
  background: #fff;
  display: flex;
  height: 100vh;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100vw;
  z-index: 10;
}

.hidden {
  display: none;
}
