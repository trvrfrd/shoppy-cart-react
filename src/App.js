import React, { Component, Fragment } from 'react';
import { hot } from 'react-hot-loader';

import Catalog from './components/Catalog';
import Cart from './components/Cart';
import CartPreview from './components/CartPreview';

// import './App.css';

class App extends Component {
  state = {
    cart: [],
    inventory: [],
    showCart: false
  }

  componentDidMount() {
    import('./inventory.json')
      .then(inventory => this.setState({ inventory: inventory.cakes }));
  }

  addItemToCart = newItem => {
    this.setState(({ cart }) => {
      const existingIdx = cart.findIndex(item => item.type === newItem.type);
      if (existingIdx !== -1) {
        cart = [...cart];
        cart[existingIdx].quantity += 1;
      } else {
        newItem = Object.assign({}, newItem, { quantity: 1 });
        cart.push(newItem);
      }

      return { cart };
    });
  }

  removeItemFromCartAt = idx => {
    this.setState(({ cart }) => {
      cart = [...cart];
      cart.splice(idx, 1);
      return { cart };
    });
  }

  clearCart = () => this.setState({ cart: [] });

  showCart = () => this.setState({ showCart: true });
  hideCart = () => this.setState({ showCart: false });

  addRandomItem = () => {
    const letters = 'abcdefghijklmnopqrstuvwxyz';
    const randomChar = letters[Math.floor(Math.random() * 26)];
    const item = {
      type: randomChar,
      price: Number((Math.random() * 10).toFixed(2))
    }
    this.addItemToCart(item);
  }

  render() {
    return (
      <Fragment>
        <header>
          <h1>Let Them Eat Cake</h1>
        </header>
        <main>
          <Catalog
            className="catalog"
            inventory={this.state.inventory}
            onAddProductToCart={this.addItemToCart}
          />
        </main>
        <aside className="sidebar">
          <CartPreview
            items={this.state.cart}
            onView={this.showCart}
          />
        </aside>
        <div style={{ display: this.state.showCart ? 'inherit' : 'none' }}>
          <Cart
            items={this.state.cart}
            onClear={this.clearCart}
            onRemoveItem={this.removeItemFromCartAt}
            onClose={this.hideCart}
          />
        </div>
      </Fragment>
    )
  }
}

export default hot(module)(App);
