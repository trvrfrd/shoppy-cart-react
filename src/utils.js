export function formatUSD(num) {
  if (isNaN(num)) throw new TypeError(`${num} is not a number`);

  num = Number(num);
  return `$${num.toFixed(2)}`;
}

export function titleCase(str) {
  return str.split(' ').map(capitalize).join(' ');
}

export function capitalize(str) {
  return str[0].toUpperCase() + str.substring(1);
}
