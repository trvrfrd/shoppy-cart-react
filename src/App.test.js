import React from 'react';
import { mount} from 'enzyme';
import TestRenderer from 'react-test-renderer';
import App from './App';

import { formatUSD, capitalize, titleCase } from './utils';

import { findByTestName } from '../test/helpers';

import { fakeItems } from '../test/mocks';
const fakeInventory = {
  // App expects array under the key cakes: which is probably dumb
  cakes: fakeItems
}

const fakeLoader = () => Promise.resolve(fakeInventory);


it('matches snapshot', async () => {
  let wrapper = await TestRenderer.create(<App loadInventory={fakeLoader} />)
  expect(wrapper).toMatchSnapshot();
});

describe('scenarios', () => {
  let wrapper;
  // I am throwing in a million async/awaits because I don't know what I"m doing
  beforeEach(async () => {
    wrapper = await mount(<App loadInventory={fakeLoader} />);
    await wrapper.update();
  });
  afterEach(async () => await wrapper.unmount());

  test('Catalog shows inventory of Products', async () => {
    const products = findByTestName(wrapper, 'product');
    expect(products.length).toBe(fakeItems.length);

    const product = products.at(0);
    expect(product.text()).toMatch(titleCase(fakeItems[0].type));
    expect(product.text()).toMatch(capitalize(fakeItems[0].description));
    expect(product.text()).toMatch(formatUSD(fakeItems[0].price));
  });

  test('Cart can be viewed and closed and starts out empty', async () => {
    const cartPreview = findByTestName(wrapper, 'cart-preview');
    expect(cartPreview.text()).toMatch('0 items');

    const viewCart = findByTestName(wrapper, 'view-cart');
    viewCart.simulate('click');

    const cart = findByTestName(wrapper, 'cart');
    expect(cart.length).toBe(1);
    expect(cart.text()).toMatch('Your cart is empty.');

    const closeCart = findByTestName(cart, 'close-cart');
    closeCart.simulate('click');
    expect(findByTestName(wrapper, 'cart').length).toBe(0);
  });

  // this is doing too much but oh well
  test('add various combinations of items to cart', async () => {
    const addToCartButtons = findByTestName(wrapper, 'add-to-cart');
    expect(addToCartButtons.length).toBe(fakeItems.length);
    addToCartButtons.at(0).simulate('click');

    const cartPreview = findByTestName(wrapper, 'cart-preview');
    expect(cartPreview.text()).toMatch('1 items');

    const viewCart = findByTestName(wrapper, 'view-cart');
    viewCart.simulate('click');

    const cart = findByTestName(wrapper, 'cart');
    expect(cart.text()).toMatch(titleCase(fakeItems[0].type));

    // this kind of covers both quantity and price table columns so fuck it
    let expectedTotal = fakeItems[0].price;
    expect(cart.text()).toMatch(`Total: ${formatUSD(expectedTotal)}`);

    addToCartButtons.at(0).simulate('click');

    expect(cartPreview.text()).toMatch('2 items');
    expectedTotal = fakeItems[0].price * 2;
    expect(cart.text()).toMatch(`Total: ${formatUSD(expectedTotal)}`);

    addToCartButtons.at(1).simulate('click');

    expect(cartPreview.text()).toMatch('3 items');
    expect(cart.text()).toMatch(titleCase(fakeItems[1].type));

    expectedTotal = fakeItems[0].price * 2 + fakeItems[1].price;
    expect(cart.text()).toMatch(`Total: ${formatUSD(expectedTotal)}`);

    addToCartButtons.at(2).simulate('click');

    expect(cartPreview.text()).toMatch('4 items');
    expect(cart.text()).toMatch(titleCase(fakeItems[2].type));

    expectedTotal = fakeItems[0].price * 2 + fakeItems[1].price + fakeItems[2].price;
    expect(cart.text()).toMatch(`Total: ${formatUSD(expectedTotal)}`);
  });

  test('remove an item from cart', async () => {
    findByTestName(wrapper, 'add-to-cart').at(0).simulate('click');

    const cartPreview = findByTestName(wrapper, 'cart-preview')
    expect(cartPreview.text()).toMatch('1 items');

    findByTestName(wrapper, 'view-cart').simulate('click');

    const cart = findByTestName(wrapper, 'cart');
    findByTestName(cart, 'remove-item').simulate('click');

    expect(cart.text()).toMatch('Your cart is empty');
    expect(cartPreview.text()).toMatch('0 items');
  });

  test('clear all items from cart (also closes cart', async () => {
    findByTestName(wrapper, 'add-to-cart').at(0).simulate('click');
    findByTestName(wrapper, 'add-to-cart').at(1).simulate('click');

    const cartPreview = findByTestName(wrapper, 'cart-preview')
    expect(cartPreview.text()).toMatch('2 items');

    findByTestName(cartPreview, 'view-cart').simulate('click');

    const cart = findByTestName(wrapper, 'cart');
    findByTestName(cart, 'clear-cart').simulate('click');

    expect(cartPreview.text()).toMatch('0 items');
    expect(findByTestName(wrapper, 'cart').length).toBe(0);
  });
});
