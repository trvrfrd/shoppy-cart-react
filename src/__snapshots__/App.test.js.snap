// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
Array [
  <header>
    <h1>
      Let Them Eat Cake
    </h1>
  </header>,
  <main>
    <div
      className="wrapper"
    >
      <ul
        className="list"
      >
        <li>
          <article
            className="product"
            data-test-name="product"
          >
            <h2
              className="name"
            >
              Product A
            </h2>
            <p
              className="description"
            >
              Amazing product
            </p>
            <p
              className="price"
            >
              $1.00
            </p>
            <div
              className="buttonWrapper"
            >
              <button
                className="button"
                data-test-name="add-to-cart"
                onClick={[Function]}
              >
                Add to Cart
              </button>
            </div>
          </article>
        </li>
        <li>
          <article
            className="product"
            data-test-name="product"
          >
            <h2
              className="name"
            >
              Product B
            </h2>
            <p
              className="description"
            >
              Beautiful product
            </p>
            <p
              className="price"
            >
              $1.50
            </p>
            <div
              className="buttonWrapper"
            >
              <button
                className="button"
                data-test-name="add-to-cart"
                onClick={[Function]}
              >
                Add to Cart
              </button>
            </div>
          </article>
        </li>
        <li>
          <article
            className="product"
            data-test-name="product"
          >
            <h2
              className="name"
            >
              Product C
            </h2>
            <p
              className="description"
            >
              I love this product more than i love my wife
            </p>
            <p
              className="price"
            >
              $2.00
            </p>
            <div
              className="buttonWrapper"
            >
              <button
                className="button"
                data-test-name="add-to-cart"
                onClick={[Function]}
              >
                Add to Cart
              </button>
            </div>
          </article>
        </li>
      </ul>
    </div>
    <aside>
      <div
        className="wrapper"
        data-test-name="cart-preview"
      >
        <button
          className="button"
          data-test-name="view-cart"
          onClick={[Function]}
        >
          View Cart
        </button>
        <p>
          (
          0
           items)
        </p>
      </div>
    </aside>
  </main>,
]
`;
